frontend/src/stimulus/controllers/pattern-autocompleter.controller.ts

```diff
@@ -0,0 +1,308 @@
+    }
+
+    // close the suggestions
+    if (event.key === 'Escape' || event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      this.clearSuggestionsFilter();
+      this.hide(this.suggestionsTarget);
+    }
+
+    // update cursor
+    this.setRange();
+  }
+
+  input_change() {
+    // browsers insert a `<br>` tag on empty contenteditable elements so we need to cleanup
+    if (this.contentTarget.innerHTML === '<br>') {
+      this.contentTarget.innerHTML = ' ';
+    }
+
+    this.ensureSpacesAround();
+
+    // show suggestions for the current word
+    const word = this.currentWord();
+    if (word && word.length > 0) {
+      this.filterSuggestions(word);
+      this.selectSuggestionAt(0);
+      this.show(this.suggestionsTarget);
+    } else {
+      this.clearSuggestionsFilter();
+      this.hide(this.suggestionsTarget);
+    }
+
+    // update cursor
+    this.setRange();
+  }
+
+  input_mouseup() {
+    this.setRange();
+  }
+
+  input_focus() {
+    this.setRange();
+  }
+
+  input_blur() {
+    this.updateFormInputValue();
+    this.hide(this.suggestionsTarget);
+  }
+
+  // Autocomplete events
+  suggestions_select(event:PointerEvent) {
+    const target = event.currentTarget as HTMLElement;
+
+    if (target) {
+      this.insertToken(this.createToken(target.dataset.value!));
+      this.clearSuggestionsFilter();
+    }
+  }
+
+  suggestions_toggle() {
+    this.clearSuggestionsFilter();
+    if (this.suggestionsTarget.getAttribute('hidden')) {
+      this.show(this.suggestionsTarget);
+    } else {
+      this.hide(this.suggestionsTarget);
+    }
+  }
+
+  // Token events
+  remove_token(event:PointerEvent) {
+    const target = event.currentTarget as HTMLElement;
+
+    if (target) {
+      const tokenElement = target.closest('[data-role="token"]');
+      if (tokenElement) {
+        tokenElement.remove();
+      }
+
+      this.updateFormInputValue();
+    }
+  }
+
+  // internal methods
+  private updateFormInputValue():void {
+    this.formInputTarget.value = this.toBlueprint();
+  }
+
+  private ensureSpacesAround():void {
+    if (this.contentTarget.innerHTML.startsWith('<')) {
+      this.contentTarget.insertBefore(document.createTextNode(' '), this.contentTarget.children[0]);
+    }
+    if (this.contentTarget.innerHTML.endsWith('>')) {
+      this.contentTarget.appendChild(document.createTextNode(' '));
+    }
+  }
+
+  private setRange():void {
+    const selection = document.getSelection();
+    if (selection?.rangeCount) {
+      const range = selection.getRangeAt(0);
+      if (range.startContainer.parentNode === this.contentTarget) {
+        this.currentRange = range;
+      }
+    }
+  }
+
+  private insertToken(tokenElement:HTMLElement) {
+    if (this.currentRange) {
+      const targetNode = this.currentRange.startContainer;
+      const targetOffset = this.currentRange.startOffset;
+
+      let pos = targetOffset - 1;
+      while (pos > -1 && targetNode.textContent?.charAt(pos) !== ' ') { pos-=1; }
```

Comment:
Mh, so maybe it's because I spent the past 10 minutes staring at the [Wikipedia page about whitespace](https://en.wikipedia.org/wiki/Whitespace_character), but seeing this code I am now wondering, whether we should not match for the space character and instead match for any kind of unicode whitespace.

I.e.

```suggestion
      while (pos > -1 && !isWhitespace(targetNode.textContent?.charAt(pos))) { pos-=1; }
```

where I leave the implementation of `isWhitespace` to the imagination of the reader.

Reply:
I found two implementations interesting :sweat_smile: 

```javascript
/\s/.test(ch)
// or
str.trim() === ''
```

This is fun~